src/backend/distributed/utils/citus_safe_lib.c

```diff
@@ -0,0 +1,287 @@
+
+	if (str == endptr)
+	{
+		ereport(ERROR, (errmsg("Error parsing %s as int64, no digits found\n", str)));
+	}
+	else if (errno == ERANGE && number == LONG_MIN)
+	{
+		ereport(ERROR, (errmsg("Error parsing %s as int64, underflow occured\n", str)));
+	}
+	else if (errno == ERANGE && number == LONG_MAX)
+	{
+		ereport(ERROR, (errmsg("Error parsing %s as int64, overflow occured\n", str)));
+	}
+	else if (errno == EINVAL)
+	{
+		ereport(ERROR, (errmsg(
+							"Error parsing %s as int64, base contains unsupported value\n",
+							str)));
+	}
+	else if (errno != 0 && number == 0)
+	{
+		int err = errno;
+		ereport(ERROR, (errmsg("Error parsing %s as int64, errno %d\n", str, err)));
+	}
+	else if (errno == 0 && str && *endptr != '\0')
+	{
+		ereport(ERROR, (errmsg(
+							"Error parsing %s as int64, aditional characters remain after int64\n",
+							str)));
+	}
+	return number;
+}
+
+
+/*
+ * SafeStringToUint64 converts a string containing a number to a uint64. When it
+ * fails it calls ereport.
+ *
+ * The different error cases are inspired by
+ * https://stackoverflow.com/a/26083517/2570866
+ */
+uint64
+SafeStringToUint64(const char *str)
+{
+	char *endptr;
+	errno = 0;
+	uint64 number = strtoul(str, &endptr, 10);
+
+	if (str == endptr)
+	{
+		ereport(ERROR, (errmsg("Error parsing %s as uint64, no digits found\n", str)));
+	}
+	else if (errno == ERANGE && number == LONG_MIN)
+	{
+		ereport(ERROR, (errmsg("Error parsing %s as uint64, underflow occured\n", str)));
+	}
+	else if (errno == ERANGE && number == LONG_MAX)
+	{
+		ereport(ERROR, (errmsg("Error parsing %s as uint64, overflow occured\n", str)));
+	}
+	else if (errno == EINVAL)
+	{
+		ereport(ERROR, (errmsg(
+							"Error parsing %s as uint64, base contains unsupported value\n",
+							str)));
+	}
+	else if (errno != 0 && number == 0)
+	{
+		int err = errno;
+		ereport(ERROR, (errmsg("Error parsing %s as uint64, errno %d\n", str, err)));
+	}
+	else if (errno == 0 && str && *endptr != '\0')
+	{
+		ereport(ERROR, (errmsg(
+							"Error parsing %s as uint64, aditional characters remain after uint64\n",
+							str)));
+	}
+	return number;
+}
+
+
+/*
+ * SafeQsort is the non reentrant version of qsort (qsort vs qsort_r), but it
+ * does the input checks required for qsort_s:
+ *  1. count or size is greater than RSIZE_MAX
+ *  2. ptr or comp is a null pointer (unless count is zero)
+ * source: https://en.cppreference.com/w/c/algorithm/qsort
+ *
+ * When it hits these errors it calls the ereport_constraint_handler.
+ *
+ * NOTE: this functions calls pg_qsort instead of stdlib qsort.
+ */
+void
+SafeQsort(void *ptr, rsize_t count, rsize_t size,
+		  int (*comp)(const void *, const void *))
+{
+	if (count > RSIZE_MAX_MEM)
+	{
+		ereport_constraint_handler("SafeQsort: count exceeds max",
+								   NULL, ESLEMAX);
+	}
+
+	if (size > RSIZE_MAX_MEM)
+	{
+		ereport_constraint_handler("SafeQsort: size exceeds max",
+								   NULL, ESLEMAX);
+	}
+	if (size != 0)
```

Comment:
Do we care about negative sizes?

Reply:
That's handled by the type system, since `rsize_t` is an alias for `size_t` which is unsigned: https://en.cppreference.com/w/c/types/size_t